### PR TITLE
chore(material): migrate most components to TypeScript.

### PR DESCRIPTION
**Pull request by jelbourn**
All components migrated to TypeScript except for the dialog, which is blocked on #2127, and Tabs which are pending in #1977.

/cc @mprobst @alexeagle

### CODE REVIEW

File: modules/angular2_material/src/components/input/input.ts [hunk 1]
```diff
@@ -10,11 +8,56 @@ import {ObservableWrapper, EventEmitter} from 'angular2/src/facade/async';
...
+
+  // Whether the input inside of this container has a non-empty value.
+  inputHasValue: boolean;
+
+  // Whether the input inside of this container has focus.
+  inputHasFocus: boolean;
+
+  constructor(@Attribute('id') id: string) {
+    this._input = null;
+    this.inputHasValue = false;
```
Comment (vicb) [inline modules/angular2_material/src/components/input/input.ts]: move the init value to the declaration ?

File: modules/examples/src/material/input/index.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import {bootstrap, Component, View} from 'angular2/angular2';
+import {MdInputContainer, MdInput} from 'angular2_material/src/components/input/input';
+import {UrlResolver} from 'angular2/src/services/url_resolver';
+import {commonDemoSetup, DemoUrlResolver} from '../demo_common';
+import {bind} from 'angular2/di';
+
+@Component({selector: 'demo-app'})
+@View({templateUrl: './demo_app.html', directives: [MdInputContainer, MdInput]})
+class DemoApp {
+  constructor() {}
```
Comment (vicb) [inline modules/examples/src/material/input/index.ts]: do you need empty ctor (also seen somewhere above)

File: modules/angular2_material/src/components/input/input.ts (hunk 1 above)
Comment (jelbourn) [inline modules/angular2_material/src/components/input/input.ts]: I would rather make minimal changes in this PR to get things working and worry about idiomatic TS later.

**Comment by vicb** [comment]
A few questions that I also had when converting to TS, do we have guidelines on:
- explicit initialization in ctor (`this.prop = val;`) vs initialization at decl (`prop: type = val;`)
- `List<Type>` vs `Type[]`

File: modules/angular2_material/src/components/input/input.ts (hunk 1 above)
Comment (vicb) [inline modules/angular2_material/src/components/input/input.ts]: makes sense. I was also asking to know if we have defined a best practice for those.

Comment (jelbourn) [inline modules/angular2_material/src/components/input/input.ts]: I don't think we've really talked about it yet. Would be a good thing to do soon.

**Comment by vicb** [comment]
Oh... LGTM

_(OT why do material use a different clang-format style ?)_

**Comment by jelbourn** [comment]
@vicb the components will eventually be their own library / repo, and I hate the single-lining of functions and blocks.

**Comment by jelbourn** [comment]
Closed via 0f3a8f369af447e52c686fe09435c8ca14e8e19c

File: modules/angular2_material/.clang-format [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Language:        JavaScript
+BasedOnStyle:    Google
+ColumnLimit:     100
+
+TabWidth:                2
+ContinuationIndentWidth: 4
+MaxEmptyLinesToKeep    : 2
+
+AllowShortBlocksOnASingleLine:              false
```
Comment (alexeagle) [inline modules/angular2_material/.clang-format]: I prefer these as well. Maybe we should consider applying to all of Angular. The delta would be pretty big but maybe worth it? @mprobst ?

Comment (mprobst) [inline modules/angular2_material/.clang-format]: I'm large apathetic to this. I generally like anything that saves vertical
space (in particular for object literals, and there in particular in
named-arguments style calls). But I don't feel too strongly about it, I
just want automated formatting that works well.

>

**Pull request closed without merging**